### ISSUE TITLE
Computing fold is slow and it blocks vim UI

### ISSUE THREAD

**Issue by wookayin** [abarker/cyfolds#1]
On a very large file (say ~1000 lines but should be a normal size), the vim UI blocks during loading of the file (i.e. entering the buffer).

Profiling gives:

```vim
:profile start profile.log | profile func * | profile file * | e python_file.py | profile pause
```

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
  636              0.701612  GetPythonFoldViaCython()
  636              0.678712  provider#python3#Call()
    2   0.031522   0.031349  <SNR>114_request()
  184   0.043075   0.026745  23()
```

We would desire cyfold's computation runs in a background (using async features as other plugins such as ale, coc.nvim, deoplete do) so as to not block the UI. It would be non-trivial to implement though.

**Comment by abarker** [comment]
Thanks for the report, but I'll need more information to reproduce it.  On Ubuntu 18.04 with Vim 8.0 on an older Dell laptop I'm getting these numbers for a file with ~10,000 lines:

```
FUNCTIONS SORTED ON SELF TIME
count  total (s)   self (s)  function
 9667              0.540751  GetPythonFoldViaCython()
```

That seems pretty responsive from a UI perspective, especially for smaller files.

The `provider#python3#Call()` line makes me think you're using Neovim, which I have not tested Cyfolds on.  If so, the slowness might or might not be related to this bug: 
https://github.com/neovim/neovim/issues/7063

Some additional information that would be helpful:

* Version of Vim or Neovim
* Operating system and computer being run on
* Compiler used to compile the Cython code

**Comment by wookayin** [comment]
I checked with vanilla vim and it is indeed super fast. So it seems to be an issue of neovim as you guessed.

- Vim: 8.1 (+python3)
- Nvim: v0.4.3 release
- Linux x64 (Debian/Ubuntu)
- gcc 9.2.1

Will test more on my extra free cycle.

**Comment by abarker** [comment]
After a discussion of the issue in the Neovim issue thread listed above I've refactored some of the code.  The code for passing values between Vim and Python has been modified to assign a Vim list of all the computed foldlevels on the first call in a fold update.  Values are then returned from that Vim list for all the calls to `foldexpr`.  This reduces the number of Vim calls to `python3` which incurs significant RPC overhead in Neovim.

The code is still not Vim 8 fast on Neovim, but it should be a lot faster and much more usable.

**Comment by wookayin** [comment]
After your fixes (cc9a6702735af89db15942074ee27a5244b771ca), it is significantly faster and I don't experience any big troubles. It is very usable, thank you. I think this issue might be closed for now (please feel free to do so), but I'll leave it up to you as you may want to add a bit more optimization.
